src/hotspot/share/nmt/vmatree.hpp

```diff
@@ -0,0 +1,375 @@
+         !Metadata::equals(stB.metadata, Metadata{})) // The metadata was changed from empty earlier
+        ) {
+      tree.upsert(B, stB);
+    }
+
+    // Finally, we need to:
+    // 1. Perform summary accounting.
+    // 2. Delete all nodes between (A, B]. Including B in the case of a noop.
+
+    if (to_be_deleted_inbetween_a_b.length() == 0
+        && LEQ_A_found && GEQ_B_found
+        && GEQ_B.address >= B) {
+      // We have smashed a hole in an existing region (or replaced it entirely).
+      // LEQ_A - A - B - GEQ_B
+      auto& rescom = diff.flag[NMTUtil::flag_to_index(LEQ_A.flag())];
+      if (LEQ_A.state.out == InOut::Reserved) {
+        rescom.reserve -= B - A;
+      } else if (LEQ_A.state.out == InOut::Committed) {
+        rescom.commit -= B - A;
+        rescom.reserve -= B - A;
+      }
+    }
+
+    AddressState prev = {A, stA}; // stA is just filler
+    MEMFLAGS flag_in = LEQ_A.flag();
+    while (to_be_deleted_inbetween_a_b.length() > 0) {
+      const AddressState delete_me = to_be_deleted_inbetween_a_b.top();
+      to_be_deleted_inbetween_a_b.pop();
+      tree.remove(delete_me.address);
+      auto& rescom = diff.flag[NMTUtil::flag_to_index(flag_in)];
+      if (delete_me.state.in == InOut::Reserved) {
+        rescom.reserve -= delete_me.address - prev.address;
+      } else if (delete_me.state.in == InOut::Committed) {
+        rescom.commit -= delete_me.address - prev.address;
+        rescom.reserve -= delete_me.address - prev.address;
+      }
+      prev = delete_me;
+      flag_in = delete_me.flag();
+    }
+    if (prev.address != A &&
+        prev.state.out != InOut::Released &&
+        GEQ_B.state.in != InOut::Released) {
+      // There was some node inside of (A, B) and it is connected to GEQ_B
+      // A - prev - B - GEQ_B
+      // It might be that prev.address == B == GEQ_B.address, this is fine.
+      if (prev.state.out == InOut::Reserved) {
+        auto& rescom = diff.flag[NMTUtil::flag_to_index(prev.flag())];
+        rescom.reserve -= B - prev.address;
+      } else if (prev.state.out == InOut::Committed) {
+        auto& rescom = diff.flag[NMTUtil::flag_to_index(prev.flag())];
+        rescom.commit -= B - prev.address;
+        rescom.reserve -= B - prev.address;
+      }
+    }
+
+    // Finally, we can register the new region [A, B)'s summary data.
+    auto& rescom = diff.flag[NMTUtil::flag_to_index(metadata.flag)];
+    if (state == InOut::Reserved) {
+      rescom.reserve += B - A;
+    } else if (state == InOut::Committed) {
+      rescom.commit += B - A;
+      rescom.reserve += B - A;
+    }
+    return diff;
+  }
+
+  static Metadata identity_merge(Metadata& merge_into, const Metadata& existent) {
+    return merge_into;
+  }
+
+  SummaryDiff reserve_mapping(size_t from, size_t sz, Metadata& metadata) {
+    return register_mapping(from, from + sz, InOut::Reserved, metadata, identity_merge);
+  }
+
+  SummaryDiff commit_mapping(size_t from, size_t sz, Metadata& metadata) {
```

Comment:
So, the `Merge` template parameter only tries to solve the problem that callers don't specify MEMFLAGS on commit or uncommit, right?

We can
- rethink that; require all callers of commit/uncommit to provide MEMFLAGS too. I know I have been arguing against it, but I like the simpler implementation of NMT. And arguably, it closer reflects `mmap` realities, where reserving and committing don't have the same roles as in hotspot.
- don't make merging a topic for the tree. If we want the MEMFLAGS from prior reservations to carry over, just search the tree prior to registering the mapping. Since we need a "tell me the region this pointer points to" functionality anyway, just let's reuse that one. 
- or, at least, revert to a simple bool or enum that describes merging behavior. Keep it simple and stupid.

Reply:
Right, merging isn't even a complete solution. Consider this situation:

```c++
Tree tree;
tree.reserve_mapping(0, 50, mtNMT);
tree.reserve_mapping(50, 50, mtTest);
tree.commit_mapping(0, 100);
```

The current merge-strategy results in:

```
0    mtNMT   100
|                       |
```

While we *actually* wanted:

```
0 mtNMT  50   mtTest 100
|                 |                     |
```

Now, we happen to be safe from this behavior as our current `VirtualMemoryTracker` doesn't allow this kind of call anyway.

I really want solution 1, it just makes sense. I don't want to make that change though :-). Not right now, at least.